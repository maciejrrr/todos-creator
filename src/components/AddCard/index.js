import React, { Component } from 'react';
import uniqueId from 'lodash/uniqueId';
import styled, { css } from 'styled-components';
import { Icon } from 'antd';

import { metrics, colors } from '../../theme';

const Form = styled.form`
  padding: ${metrics.smallMargin}px;
  border-radius: ${metrics.borderRadius}px;
  background-color: ${props =>
    props.focused ? colors.cardBackground : 'transparent'};
`;

const Input = styled.input`
  background-color: ${colors.addCardBackground};
  border: 0;
  outline-width: 0;
  height: ${metrics.inputHeight}px;
  padding: ${metrics.smallMargin}px;
  color: ${colors.addCardText};
  ::placeholder {
    color: ${colors.addCardText};
  }
  ${props =>
    props.focused &&
    css`
      color: ${colors.cardText};
      ::placeholder {
        color: ${colors.cardText};
      }
    `};
`;

const Actions = styled.div`
  margin-top: ${metrics.smallMargin}px;
  display: flex;
  align-items: center;
`;

const Submit = styled.button`
  border: 0;
  border-radius: ${metrics.borderRadius}px;
  background-image: linear-gradient(${colors.gradient});
  padding: ${metrics.smallMargin}px ${metrics.doubleMargin}px;
  color: ${colors.white};
  ${props =>
    props.disabled &&
    css`
      cursor: not-allowed;
      background-image: none;
      background-color: ${colors.white};
      color: ${colors.cardText};
    `};
`;

const Cancel = styled(Icon)`
  cursor: pointer;
  font-size: ${metrics.fontSize.xl}px;
  margin-left: ${metrics.smallMargin}px;
`;

class AddCard extends Component {
  state = {
    name: '',
    focused: false,
  };

  handleSubmit = e => {
    e.preventDefault();
    this.props.addCard({
      card: {
        id: uniqueId(),
        name: this.state.name,
      },
    });
    this.handleResetForm();
  };

  handleChange = e => {
    this.setState({ name: e.target.value });
  };

  handleFocus = () => {
    this.setState({ focused: true });
  };

  handleResetForm = () => {
    this.setState({ focused: false, name: '' });
  };

  render() {
    const { focused, name } = this.state;
    return (
      <Form focused={focused} onSubmit={this.handleSubmit}>
        <Input
          focused={focused}
          placeholder="Add a card"
          onChange={this.handleChange}
          value={name}
          onFocus={this.handleFocus}
        />
        {focused && (
          <Actions>
            <Submit htmlType="submit" disabled={!name}>
              Add
            </Submit>
            <Cancel type="close" onClick={this.handleResetForm} />
          </Actions>
        )}
      </Form>
    );
  }
}

export default AddCard;
