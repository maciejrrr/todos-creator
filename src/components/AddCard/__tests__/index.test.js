import AddCard from '../index';

describe('AddCard component', () => {
  let props;
  beforeEach(() => {
    props = {
      addCard: jest.fn(),
    };
  });

  describe('snapshots', () => {
    describe('initial state', () => {
      it('renders without focused components', () => {
        const tree = renderer.create(<AddCard {...props} />).toJSON();
        expect(tree).toMatchSnapshot();
      });
    });

    describe('focused state', () => {
      it('renders focused form', () => {
        const component = renderer.create(<AddCard {...props} />);
        component.getInstance().handleFocus();
        expect(component.toJSON()).toMatchSnapshot();
      });
    });
  });

  describe('handleSubmit', () => {
    let e;
    let preventDefaultMock;
    let handleResetFormMock;
    let blurMock;
    let name;
    beforeEach(() => {
      preventDefaultMock = jest.fn();
      handleResetFormMock = jest.fn();
      blurMock = jest.fn();
      const component = shallow(<AddCard {...props} />);
      e = { preventDefault: preventDefaultMock };
      name = 'valid name';
      component.instance().handleResetForm = handleResetFormMock;
      component.instance().input = { blur: blurMock };
      component.setState({ name, focused: true });
      component.instance().handleSubmit(e);
    });

    it('calls preventDefault', () => {
      expect(preventDefaultMock).toHaveBeenCalledTimes(1);
    });

    it('calls addCard', () => {
      expect(props.addCard).toBeCalledWith({
        card: {
          id: 1,
          name,
        },
      });
    });

    it('calls handleResetForm', () => {
      expect(handleResetFormMock).toHaveBeenCalledTimes(1);
    });

    it('calls blur on input ref', () => {
      expect(blurMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('handleChange', () => {
    it('changes name state', () => {
      const component = shallow(<AddCard {...props} />);
      const value = 'test value';
      const e = { target: { value } };
      component.instance().handleChange(e);
    });
  });

  describe('handleFocus', () => {
    it('changes focused state', () => {
      const component = shallow(<AddCard {...props} />);
      component.instance().handleFocus();
      expect(component.state().focused).toBe(true);
    });
  });

  describe('handleResetForm', () => {
    let component;
    beforeEach(() => {
      component = shallow(<AddCard {...props} />);
      component.setState({ focused: true, name: 'test name' });
      component.instance().handleResetForm();
    });

    it('changes focused state', () => {
      expect(component.state().focused).toBe(false);
    });

    it('changes name state', () => {
      expect(component.state().name).toBe('');
    });
  });
});
