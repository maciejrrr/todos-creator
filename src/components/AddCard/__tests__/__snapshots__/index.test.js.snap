// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCard component snapshots focused state renders focused form 1`] = `
<form
  className="sc-bxivhb umNtg"
  onSubmit={[Function]}
>
  <input
    className="sc-ifAKCX cRsMfw"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Add a card..."
    value=""
  />
  <div
    className="sc-bdVaJa cItXoG"
  >
    <button
      className="sc-bwzfXH hGZcQI"
      disabled={true}
    >
      Add
    </button>
    <i
      className="anticon anticon-close sc-htpNat oPmsX"
      onClick={[Function]}
    />
  </div>
</form>
`;

exports[`AddCard component snapshots initial state renders without focused components 1`] = `
<form
  className="sc-bxivhb ctFYVZ"
  onSubmit={[Function]}
>
  <input
    className="sc-ifAKCX cMqBtW"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Add a card..."
    value=""
  />
</form>
`;
